### PR TITLE
bpo-21071: struct.Struct.format type is now str

### PR DESCRIPTION
**Pull request by vstinner**

### CODE REVIEW

File: Misc/NEWS [hunk 1]
```diff
@@ -291,6 +291,9 @@ Extension Modules
 Library
 -------
 
+- bpo-21071: struct.Struct.format is now a Unicode instead, instead of a bytes
```
Comment (vadmium) [inline Misc/NEWS]: Unicode instead → Unicode string
bytes string → byte string [or bytes object if you prefer]

File: Doc/library/struct.rst [hunk 2]
```diff
@@ -443,6 +443,9 @@ The :mod:`struct` module also defines the following type:
 
       The format string used to construct this Struct object.
 
+      .. versionchanged:: 3.6
+         The format string is now a Unicode string, instead of a bytes string.
```
Comment (vadmium) [inline Doc/library/struct.rst]: 3.6 → 3.7
I think either “byte string” or “bytes object” would read better.

Comment (serhiy-storchaka) [reply Doc/library/struct.rst]: Maybe use just `str` and `bytes`? There is nothing specific to the Unicode standard here.

Comment (vstinner) [reply Doc/library/struct.rst]: Fixed, I used "str" and "bytes" types.

File: Misc/NEWS (hunk 1 above)
Comment (vstinner) [reply Misc/NEWS]: Fixed

**Comment by vstinner** [comment]
@vadmium, @serhiy-storchaka: I updated my PR to take in account your remarks. Would you mind to review my PR again, please?

File: Misc/NEWS [hunk 3]
```diff
@@ -362,6 +362,9 @@ Extension Modules
 Library
 -------
 
+- bpo-21071: struct.Struct.format type is now :class:`str` instead of
```
Comment (serhiy-storchaka) [inline Misc/NEWS:377]: I think `:class:` is not needed.

File: Modules/_struct.c [hunk 4]
```diff
@@ -1957,8 +1957,8 @@ s_pack_into(PyObject *self, PyObject **args, Py_ssize_t nargs, PyObject *kwnames
 static PyObject *
 s_get_format(PyStructObject *self, void *unused)
 {
-    Py_INCREF(self->s_format);
-    return self->s_format;
+    return PyUnicode_FromStringAndSize(PyBytes_AS_STRING(self->s_format),
```
Comment (serhiy-storchaka) [inline Modules/_struct.c:1960]: Why not make `s_format` an instance of str?

**Review by serhiy-storchaka — COMMENTED** [review]
Since this change breaks backward compatibility, it should be documented in What's New.

What is the result of the discussion on Python-Dev?

**Comment by vstinner** [comment]
> What is the result of the discussion on Python-Dev?

No answer: https://mail.python.org/pipermail/python-dev/2017-March/147688.html

File: Modules/_struct.c (hunk 4 above)
Comment (vstinner) [reply Modules/_struct.c:1960]: I tried to write minimum changes. I expect that most of the code is written to work with C strings char*, not with Python Unicode strings.

**Comment by vstinner** [comment]
> Since this change breaks backward compatibility, it should be documented in What's New.

Ok, done.

The change is now documented in the changelog (Misc/NEWS), twice in What's New in Python 3.7 (struct module and Porting to Python 3.7), and in struct documentation. I don't think that you can document it more :-)

Seriously, I don't think that anyone rely on the exact type of struct.Struct.format. If someone cares, it will be trivial to notice, and also likely trivial to fix the code.

File: Doc/whatsnew/3.7.rst [hunk 5]
```diff
@@ -425,6 +431,9 @@ Changes in the Python API
   ``makedirs()``.
   (Contributed by Serhiy Storchaka in :issue:`19930`.)
 
+* The :attr:`struct.Struct.format` type is now :class:`str` instead of
+  :class:`bytes`
```
Comment (serhiy-storchaka) [inline Doc/whatsnew/3.7.rst]: Add a reference to issue.

File: Doc/whatsnew/3.7.rst [hunk 6]
```diff
@@ -232,6 +232,12 @@ New function :func:`os.register_at_fork` allows registering Python callbacks
 to be executed on a process fork.  (Contributed by Antoine Pitrou in
 :issue:`16500`.)
 
+struct
+------
+
+The :attr:`struct.Struct.format` type is now :class:`str` instead of
```
Comment (serhiy-storchaka) [inline Doc/whatsnew/3.7.rst]: Actually I'm not sure that mentioning this change here is needed. This is not a new *feature*.

**Review by serhiy-storchaka — APPROVED** [review]
LGTM, but add a reference to the issue and wait yet short time for a reaction on Python-Dev.

File: Doc/whatsnew/3.7.rst (hunk 5 above)
Comment (vstinner) [reply Doc/whatsnew/3.7.rst]: ok, done

File: Doc/whatsnew/3.7.rst (hunk 6 above)
Comment (vstinner) [reply Doc/whatsnew/3.7.rst]: ok, removed

File: Modules/_struct.c (hunk 4 above)
Comment (vadmium) [reply Modules/_struct.c:1960]: FWIW I already did this in my _format-str.patch_. I was able to get a _char*_ out of the string with _PyUnicode_AsUTF8_.

File: Lib/test/test_struct.py [hunk 7]
```diff
@@ -618,6 +618,14 @@ def test_issue29802(self):
         # Shouldn't crash.
         self.assertEqual(struct.unpack(b'b', b'a'), (b'a'[0],))
 
+    def test_format_attr(self):
+        s = struct.Struct('=i2H')
+        self.assertEqual(s.format, '=i2H')
```
Comment (berkerpeksag) [inline Lib/test/test_struct.py:623]: IMO, `self.assertIsInstance(s.format, str)` would be better to understand (for future readers) the behavior change.

**Comment by vstinner** [comment]
@serhiy-storchaka: "LGTM, but add a reference to the issue and wait yet short time for a reaction on Python-Dev."

Serhiy wrote an email to my old thread:
https://mail.python.org/pipermail/python-dev/2017-June/148360.html

@ncoghlan wrote "As long as it's noted in the "Porting to Python 3.7" section of the 3.7 What's New guide, this seems like a sensible change to me."

It's the case, I added an entry to Porting to Python 3.7. I will wait until Friday (June 23) evening.

**Comment by vstinner** [comment]
> It's the case, I added an entry to Porting to Python 3.7. I will wait until Friday (June 23) evening.

It's today. I began by rebasing my change.

**Pull request merged**